provider/sourcegraph-search/index.ts

```diff
@@ -0,0 +1,131 @@
+import type {
+    AnnotationsParams,
+    AnnotationsResult,
+    CapabilitiesParams,
+    CapabilitiesResult,
+    Item,
+    ItemsParams,
+    ItemsResult,
+    Provider,
+    ProviderSettings,
+} from '@openctx/provider'
+import { SourcegraphGraphQLAPIClient } from './graphql'
+import { isError, searchForFileChunks } from './search'
+
+const graphqlClient = new SourcegraphGraphQLAPIClient()
+
+interface ItemsParamsProposed extends ItemsParams {
+    candidate?: Item
+}
+
+interface ProviderProposed<S extends {} = ProviderSettings> extends Provider {
+    candidateItems?(params: ItemsParamsProposed, settings: S): ItemsResult | Promise<ItemsResult>
+    items?(params: ItemsParamsProposed, settings: S): ItemsResult | Promise<ItemsResult>
+}
+
+/** Used to prefix context queries to indicate a sourcegraph search */
+const QUERY_PREFIXES = ['?', 'src:']
+
+/** We always limit our searches to files and a conservatively sized result set */
+const SEARCH_QUERY_CONST = 'type:file count:10'
+
+const sourcegraphSearch: ProviderProposed = {
+    capabilities(params: CapabilitiesParams, settings: ProviderSettings): CapabilitiesResult {
+        return {
+            // empty since we don't provide any annotations.
+            selector: [],
+        }
+    },
+
+    async candidateItems(params: ItemsParamsProposed): Promise<ItemsResult> {
+        if (!params.query) {
+            return []
+        }
+
+        const query = searchQueryFromItemQuery(params.query)
+        const url = `${graphqlClient.endpoint}/search?q=${encodeURIComponent(
+            query.sourcegraphQuery
+        )}&patternType=literal`
+
+        return [
+            {
+                title: query.sourcegraphQuery,
+                url,
+                ui: { hover: { text: `search on ${graphqlClient.endpoint}` } },
+            },
+        ]
+    },
+
+    async items(params: ItemsParamsProposed, settings: ProviderSettings): Promise<ItemsResult> {
+        let query: SearchQuery | undefined = undefined
+        if (params.candidate?.url) {
+            query = searchQueryFromURL(graphqlClient.endpoint, params.candidate.url)
+        }
+        if (!query && params.query) {
+            query = searchQueryFromItemQuery(params.query)
+        }
+        if (!query) {
+            return []
+        }
```

Comment:
I treat query and candidate as only one is allowed to be set. Does that make sense in general? If so is that something we should encode in the schema?